File: Python/ByTier/C/1879_C_Make_it_Alternating.py
# Problem: CF 1879 C - Make it Alternating
# https://codeforces.com/contest/1879/problem/C

"""
Algorithm: Make it Alternating
Approach:
1. The problem requires transforming a binary string into an alternating string (0101... or 1010...) by deleting minimum number of characters.
2. Key insight: We can group consecutive identical characters and determine how many deletions are needed to break up long sequences.
3. The minimum deletions = total length - number of groups.
4. To compute number of ways:
   - Each group contributes a certain number of choices for deletion (equal to group size)
   - Total ways = product of choices for each group.
   - But we also need to consider that the deletions themselves form a sequence, so there are multiple orders of doing the same deletions.

Time Complexity: O(n) where n is total length of input string across all test cases.
Space Complexity: O(n) for storing grouped character counts.

Techniques:
- Grouping consecutive identical elements
- Combinatorial counting with modular arithmetic
- Modular multiplication for large results
"""

import sys

input = sys.stdin.readline


def inp():
    return int(input())


def inlt():
    return list(map(int, input().split()))


def insr():
    s = input()
    return list(s[: len(s) - 1])


def invr():
    return map(int, input().split())


tt = inp()
for _ in range(tt):
    s = insr()
    myList = []
    curElem = s[0]
    curCount = 0
    for c in s:
        if c != curElem:
            curElem = c
            if curCount > 1:
                myList.append(curCount)
            curCount = 1
        else:
            curCount += 1
    if curCount > 1:
        myList.append(curCount)
    numWays = 1
    deletions = sum(myList) - len(myList)
    for num in myList:
        numWays = (numWays * num) % 998244353
    for i in range(2, deletions + 1):
        numWays = (numWays * i) % 998244353
    print(deletions, numWays)


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1879/1879_C_Make_it_Alternating.cpp
// Problem: CF 1879 C - Make it Alternating
// https://codeforces.com/contest/1879/problem/C

/*
C. Make it Alternating
Algorithms/Techniques: String processing, combinatorics, counting operations to make alternating string.

Time Complexity: O(n) where n is the length of the input string s (linear pass through string and computation)
Space Complexity: O(1) - only using a few variables for computation

Task:
Given a binary string s, find:
1. Minimum number of operations to make it alternating (delete characters)
2. Number of different shortest sequences of such operations modulo 998244353

Approach:
- For each group of consecutive identical characters, we can keep one and delete the rest.
- The minimum operations = total length - number of unique groups + 1
- To compute number of ways:
  - For each group of size cnt, there are cnt choices for which element to keep
  - Multiply all such choices together
  - Then multiply by factorial of (sum of deletions) because we can delete in any order among deletion operations
*/

#include <iostream>
#include <string>

typedef long long ll;
int main() {
  const ll MOD = 998244353;
  ll t;
  std::cin >> t;
  while (t--) {
    std::string s;
    std::cin >> s;
    s += "_"; // add sentinel to simplify loop logic
    ll cnt(1), sum(0), prod(1);
    for (long p = 1; p < s.size(); p++) {
      if (s[p - 1] == s[p]) {
        ++cnt;
        continue;
      }
      // When character changes:
      // - add (cnt - 1) to sum (number of deletions in this group)
      // - multiply prod by cnt (choices for which character to keep)
      sum += (cnt - 1);
      prod = (prod * cnt) % MOD;
      cnt = 1;
    }
    // Multiply result by factorial of number of deletions
    for (int p = 2; p <= sum; p++) {
      prod = (prod * p) % MOD;
    }
    std::cout << sum << " " << prod << std::endl;
  }
}


// https://github.com/VaHiX/codeForces/